### PR TITLE
DQM: New tests and Demo

### PR DESCRIPTION
**Pull request by schneiml**
#### PR description:

This PR imports the `Demo` package, replacing the old examples, from the the `new DQMStore` branch, as well as some new tests. These changes where part of #28407  before, but make more sense as an independent PR. However, this PR now depends on #28407. [Edit: rebased now that #28407 is merged.]

- The `DQMServices/Demo` package, which contains some simple modules of each class and configuration to exercise the DQM framework, with no dependence on "real" CMSSW configurations. This replaces the old DQM examples code, which used to show types of interaction with DQM that are not supported any more (much of it was deleted in previous PRs). The demo code still relies on unusual features, since it was designed to test behaviors that need to be respected by a EDM Product driven DQM as prototyped in the "new DQMStore" branch; since the current DQM is not fully based on products, the behavior is not complete for now. 

- The old `runTheWhiteRabbit` tests are removed and replaced with completely new tests. The new tests don't cover much about the actual *correctness* of the DQM output (it turns out that in the details many things are arguably wrong in the current implementation), but it tests running the DQM framework code in many common and nonstandard situations, to see that at least nothing crashes and basic sanity checks are fulfilled. Highlights include things like ProtoBuf output (only running in HLT normally), concurrent Lumisections (well actually, they are blocked by the modules, but it will use them once possible), and a simple memory consumption check (hoping this does not blow up in IB testing...)

#### PR validation:

Only new code added here, and old tests removed. The old tests, as far as I remember, where running in the IB tests but never caught anything, even as large parts of DQM changed (unlike for example the `FwkIO` tests or the remaining `fastHadd` related tests).

The new tests are more generic, but cover many of the sanity checks I usually use during development. We could be more specific about the expected output, but I don't want to set the old, arguably buggy, behavior in stone.

### CODE REVIEW

**Comment by schneiml** [comment]
please test

should be all fine; the warnings from #28407 before should be fixed.

File: DQMServices/Demo/plugins/DemoGlobalDQMEDAnalyzer.cc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  ibook.setCurrentFolder(folder_);
+  histos.histo_ = ibook.book1D("EXAMPLE", "EXAMPLE", 10, 0., 10.);
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void DemoGlobalDQMEDAnalyzer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // The following says we do not know what parameters are allowed so do no
+  // validation
+  // Please change this to state exactly what you do use, even if it is no
+  // parameters
```
Comment (makortel) [inline DQMServices/Demo/plugins/DemoGlobalDQMEDAnalyzer.cc]: This comment does not really reflect the content below (description is explicitly set). I suppose these files are intended to be used as an examples, could you revise or remove the comment? (same for other examples below that contain the same comment)

Comment (schneiml) [reply DQMServices/Demo/plugins/DemoGlobalDQMEDAnalyzer.cc]: Yes, I should really clean up these. This is basically what falls out of `mkdqmedanalyzer` with minimal changes to do sth. useful.

Comment (makortel) [reply DQMServices/Demo/plugins/DemoGlobalDQMEDAnalyzer.cc]: Ok. I suppose the template should be changed then as well
https://github.com/cms-sw/cmssw/blob/681f8b812b3a9fbb12621353305b80264bd356ee/FWCore/Skeletons/scripts/mkTemplates/DQMEDAnalyzer/DQMEDAnalyzer.cc#L141-L147
(in the `EDAnalyzer` template the comments are consistent with the code
https://github.com/cms-sw/cmssw/blob/681f8b812b3a9fbb12621353305b80264bd356ee/FWCore/Skeletons/scripts/mkTemplates/EDAnalyzer/EDAnalyzer.cc#L138-L148
)

**Comment by schneiml** [comment]
Looks all fine on my side, but needs rebase after #28407.

Should only require DQM signature after rebase.

**Comment by tlampen** [comment]
please test

**Comment by schneiml** [comment]
Still waiting for dependencies to be merged (#28407).

**Comment by tlampen** [comment]
please test with #28407

**Comment by schneiml** [comment]
@tlampen the test is rather pointless, this PR *includes*  #28407 anyways.

**Comment by tlampen** [comment]
> @tlampen the test is rather pointless, this PR _includes_ #28407 anyways.

Ok, I see .

**Comment by tlampen** [comment]
+1

**Comment by schneiml** [comment]
Rebased to a more recent IB with some of my other PRs merged. 

I did not address @makortel 's comments regrading the skeleton files since `code-format` seems to mess up the skeleton when I change it, and I don't feel like debugging that now. The misleading comments in the new code are fixed, though.

The tests are not fully up-to-date w.r.t. #28622, but I'll leave it like that.

**Comment by schneiml** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Comment by smuzaffar** [comment]
@silviodonato , after the integration of this PR, couple of unit tests start failing for PY3 and DEVEL IBs [a]. Please note that for these IBs we use python3. 
@schneiml can you please provide a fix for these?

@silviodonato

[a] https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc820/CMSSW_11_1_DEVEL_X_2020-01-15-2300/unitTestLogs/DQMServices/Demo#/
```
+ cmsRun /data/cmsbld/jenkins/workspace/ib-run-qa/CMSSW_11_1_DEVEL_X_2020-01-15-2300/src/DQMServices/Demo/test/run_harvesters_cfg.py inputFiles=part1.root inputFiles=part2.root inputFiles=part3.root inputFiles=part4.root outfile=merged.root nomodules=True
5679:	----- Begin Fatal Exception 16-Jan-2020 03:00:37 CET-----------------------
5680:	An exception of category 'ConfigFileReadError' occurred while
5681:	   [0] Processing the python configuration file named /data/cmsbld/jenkins/workspace/ib-run-qa/CMSSW_11_1_DEVEL_X_2020-01-15-2300/src/DQMServices/Demo/test/run_harvesters_cfg.py
5682:	Exception Message:
5683:	 unknown python problem occurred.
5684:	SyntaxError: ('invalid syntax', ('/data/cmsbld/jenkins/workspace/ib-run-qa/CMSSW_11_1_DEVEL_X_2020-01-15-2300/src/DQMServices/Demo/test/run_harvesters_cfg.py', 22, 10, 'print args.inputFiles\n'))
5685:	----- End Fatal Exception -------------------------------------------------
```

https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc820/CMSSW_11_1_PY3_X_2020-01-15-2300/unitTestLogs/DQMServices/Demo#/

```
++ dqmiolistmes.py alltypes.root -r 1
126:	++ wc -l
127:	+ '[' 99 = 1 ']'
128:	status = 256
129:	
130:	---&gt; test TestDQMServicesDemo had ERRORS
```

**Comment by schneiml** [comment]
@smuzaffar I fixed some stuff in #28622 , but I guess I should properly test that on a py3 IB. Will provide a fix later.

**Pull request merged**